pfcpiface/aether.go

```diff
@@ -0,0 +1,849 @@
+	return a.bess.SendMsgToUPF(method, all, updated)
+}
+
+// Poll the given interface until an IPv4 address is found on it.
+func waitForIpConfigured(ctx context.Context, link netlink.Link) (*net.IPNet, error) {
+	ticker := time.NewTicker(time.Millisecond * 250)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			addrs, err := netlink.AddrList(link, unix.AF_INET)
+			if err != nil {
+				return nil, err
+			}
+			// TODO(max): filter addresses?
+			for _, addr := range addrs {
+				return addr.IPNet, nil
+			}
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+	}
+}
+
+func (a *aether) syncInterface(ctx context.Context, iface string) (err error) {
+	link, err := netlink.LinkByName(iface)
+	if err != nil {
+		return
+	}
+
+	// Own MAC.
+	copy(a.datapathMAC, link.Attrs().HardwareAddr)
+
+	if len(a.datapathMAC) != 6 {
+		return ErrInvalidArgumentWithReason("a.datapathMAC", a.datapathMAC, "invalid mac address")
+	}
+
+	// Fetch own IPv4 address.
+	a.ownIp, err = waitForIpConfigured(ctx, link)
+	a.ownIp.IP = a.ownIp.IP.To4() // Make sure it's in 4 byte format.
+	if a.ownIp == nil || a.ownIp.IP == nil {
+		return ErrOperationFailedWithReason("a.ownIp", "found no IPv4 address")
+	}
+	log.Tracef("found IP %v on local %v interface", a.ownIp, iface)
+
+	routes, err := netlink.RouteList(link, unix.AF_INET)
+	if err != nil {
+		return
+	}
+
+	for _, r := range routes {
+		log.Tracef("route: %+v", r)
+
+		// Bridging
+		if r.Scope == netlink.SCOPE_LINK {
+			log.Traceln("Found route with scope link:", r)
+			a.routeToFabric = r
+			break
+		}
+	}
+
+	if a.routeToFabric.Dst == nil {
+		return ErrOperationFailedWithReason("syncInterface", "found no route to fabric")
+	}
+
+	// Fetch default route and gateway from alternative routing table.
+	altRoutes, err := netlink.RouteListFiltered(
+		unix.AF_INET,
+		&netlink.Route{Table: bessRoutingTable},
+		netlink.RT_FILTER_TABLE)
+	if err != nil {
+		return
+	}
+	for _, r := range altRoutes {
+		if r.Gw != nil {
+			log.Traceln("Found default route with gateway IP:", r)
+			copy(a.gatewayIP, r.Gw.To4())
+			break
+		}
+	}
+
+	if a.gatewayIP == nil {
+		return ErrOperationFailedWithReason("syncInterface", "found no gateway to fabric")
+	}
+
+	a.gatewayMAC, err = a.resolveNeighbor(ctx, link.Attrs().Index, a.gatewayIP)
+	if err != nil {
+		return
+	}
+
+	log.Traceln("resolved neighbor mac", a.gatewayMAC)
+
+	return a.setupRoutingRules()
+}
+
+func (a *aether) startInterfaceWatchTask(iface string) (err error) {
+	link, err := netlink.LinkByName(iface)
+	if err != nil {
+		return
+	}
+
+	updates := make(chan netlink.AddrUpdate)
+
+	if err = netlink.AddrSubscribe(updates, a.addrUpdatesDone); err != nil {
+		return
+	}
+
+	go a.interfaceWatchTask(link, updates)
+
+	return
+}
+
+func (a *aether) interfaceWatchTask(link netlink.Link, updates <-chan netlink.AddrUpdate) {
+	for {
+		update, ok := <-updates
+		if !ok {
+			log.Infof("Address update subscriber channel for %v interface closed", link.Attrs().Name)
```

Comment:
Do we want to print a warning for some thing that's expected to happen as part of a normal shutdown flow?
Because the netlink lib will close the channel on shutdown:
https://github.com/vishvananda/netlink/blob/main/addr_linux.go#L362